wgengine/monitor/monitor_windows.go

```diff
@@ -51,229 +33,136 @@ type messageOrError struct {
-		}
-	} else {
-		switch eventNum {
-		case 1:
-			eventStr = "addr"
-		case 2:
-			eventStr = "route"
-		default:
-			eventStr = fmt.Sprintf("%d [unexpected]", eventNum)
+	for {
+		select {
+		case me := <-m.messagec:
+			return me.message, me.error
+		case <-pollTicker.C:
+			newState, err := interfaces.GetState()
+			if err != nil {
+				// Log and ignore
+				m.logf("interfaces.GetState error: %v", err)
+			} else if !prevState.Equal(newState) {
+				m.logf("interface state changed (on poll)")
+				return pollStateChangedMessage{}, nil
+			}
 		}
 	}
-	m.logf("got windows change event after %v: evt=%s", d, eventStr)
-
-	m.mu.Lock()
-	{
-		m.lastNetChange = time.Now()
-
-		// Something changed, so assume Windows is about to
-		// discover its new proxy settings from WPAD, which
-		// seems to take a bit. Poll heavily for awhile.
-		m.inFastPoll = true
-		m.pollTicker.Reset(pollIntervalFast)
-	}
-	m.mu.Unlock()
-
-	return unspecifiedMessage{}, nil
 }
 
-func notifyAddrChange() (h windows.Handle, o *windows.Overlapped, cancel func(), err error) {
-	return callNotifyProc(notifyAddrChangeProc)
+// unicastAddressChanged is the callback we register with Windows to call when unicast address changes.
+func (m *winMon) unicastAddressChanged(_ winipcfg.MibNotificationType, _ *winipcfg.MibUnicastIPAddressRow) {
+	m.somethingChanged()
 }
 
-func notifyRouteChange() (h windows.Handle, o *windows.Overlapped, cancel func(), err error) {
-	return callNotifyProc(notifyRouteChangeProc)
+// routeChanged is the callback we register with Windows to call when route changes.
+func (m *winMon) routeChanged(_ winipcfg.MibNotificationType, _ *winipcfg.MibIPforwardRow2) {
+	m.somethingChanged()
 }
 
-func callNotifyProc(p *syscall.LazyProc) (h windows.Handle, o *windows.Overlapped, cancel func(), err error) {
-	o = new(windows.Overlapped)
+// somethingChanged gets called from OS callbacks whenever address or route changes.
+func (m *winMon) somethingChanged() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
 
-	// TODO(bradfitz): understand why this if-false code doesn't
-	// work, even though the docs online suggest we should pass an
-	// event in the overlapped.Hevent field.
-	// The docs at
-	// https://docs.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-overlapped
-	// says that o.HEvent can be zero, though, which seems to work.
-	// Note that the returned windows.Handle returns the same value for both
-	// notifyAddrChange and notifyRouteChange, which is why our caller needs
-	// to look at the returned Overlapped's Internal field to see which case
-	// fired. That's also worth understanding more.
-	// See crawshaw's comment at https://github.com/tailscale/tailscale/pull/944#discussion_r526469186
-	// too.
-	if false {
-		evt, err := windows.CreateEvent(nil, 0, 0, nil)
-		if err != nil {
-			return 0, nil, nil, err
+	if m.messagec != nil {
+		m.messagec <- messageOrError{
```

Comment:
we don't want to block forever here. This should probably be a select with one case sending and another reading on a closed chan.

Reply:
As far as I know it cannot lock there because it's in the locked code (`m.messagec` isn't `nil`), and I didn't limit the `chan` to only one item.

Reply:
It will block if there is no reader.

I'd add a select.

Reply:
ok